RPC: panic when calling getstakeversions with a negative number.

The following panic was reported by chapjc on slack:

``` 
panic: runtime error: makeslice: cap out of range goroutine

1674 [running]:
github.com/decred/dcrd/blockchain.(*BlockChain).GetStakeVersions(0xc420130000, 0xc42154e0c0, 0xc4fffffe8e, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/j/go/src/github.com/decred/dcrd/blockchain/chain.go:316 +0x127
main.handleGetStakeVersions(0xc42000b100, 0xa5b780, 0xc42154e040, 0x0, 0xc4204119d0, 0xc42168d201, 0xc42009ef70, 0xc4200f0090)
        /home/j/go/src/github.com/decred/dcrd/rpcserver.go:3949 +0xb6
main.(*rpcServer).standardCmdResult(0xc42000b100, 0xc421a2fb00, 0x0, 0x10, 0x105d360, 0x0, 0x0)
        /home/j/go/src/github.com/decred/dcrd/rpcserver.go:6213 +0x97
main.(*wsClient).handleMessage(0xc421a58f30, 0xc421d53800, 0x88, 0x600)
        /home/j/go/src/github.com/decred/dcrd/rpcwebsocket.go:1410 +0x89b
main.(*wsClient).inHandler(0xc421a58f30)
        /home/j/go/src/github.com/decred/dcrd/rpcwebsocket.go:1455 +0xac
created by main.(*wsClient).Start
        /home/j/go/src/github.com/decred/dcrd/rpcwebsocket.go:1740 +0x107
```

I have traced this down to being caused by neither the RPC handler, nor the `GetStakeVersions` checking the input arguments for correctness.

--- a/blockchain/chain.go
+++ b/blockchain/chain.go
@@ -305,8 +305,22 @@ type StakeVersions struct {
 // GetStakeVersions returns a cooked array of StakeVersions.  We do this in
 // order to not bloat memory by returning raw blocks.
 func (b *BlockChain) GetStakeVersions(hash *chainhash.Hash, count int32) ([]StakeVersions, error) {
+	// Nothing to do if no count requested.
+	if count == 0 {
+		return nil, nil
+	}
+
+	if count < 0 {
+		return nil, fmt.Errorf("count must not be less than zero - "+
+			"got %d", count)
+	}
+
 	b.chainLock.Lock()
 	defer b.chainLock.Unlock()
+
+	if count > int32(b.bestNode.height) {
+		count = int32(b.bestNode.height)
+	}
 
 	startNode, err := b.findNode(hash, 0)
 	if err != nil {
@@ -315,7 +329,7 @@ func (b *BlockChain) GetStakeVersions(hash *chainhash.Hash, count int32) ([]Stak
 
 	result := make([]StakeVersions, 0, count)
 	prevNode := startNode
-	for i := int32(0); i < count; i++ {
+	for i := int32(0); prevNode != nil && i < count; i++ {
 		sv := StakeVersions{
 			Hash:         prevNode.hash,
 			Height:       prevNode.height,
@@ -329,10 +343,6 @@ func (b *BlockChain) GetStakeVersions(hash *chainhash.Hash, count int32) ([]Stak
 		prevNode, err = b.getPrevNodeFromNode(prevNode)
 		if err != nil {
 			return nil, err
-		}
-		if prevNode == nil {
-			// Just return what we did find.
-			break
 		}
 	}
 

--- a/rpcserver.go
+++ b/rpcserver.go
@@ -3936,14 +3936,17 @@ func handleGetStakeVersionInfo(s *rpcServer, cmd interface{}, closeChan <-chan s
 
 // handleGetStakeVersions implements the getstakeversions command.
 func handleGetStakeVersions(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (interface{}, error) {
-	c, ok := cmd.(*dcrjson.GetStakeVersionsCmd)
-	if !ok {
-		return nil, internalRPCError("Invalid command",
-			"handleGetStakeVersions")
-	}
+	c := cmd.(*dcrjson.GetStakeVersionsCmd)
+
 	hash, err := chainhash.NewHashFromStr(c.Hash)
 	if err != nil {
-		return nil, err
+		return nil, rpcDecodeHexError(c.Hash)
+	}
+	if c.Count <= 0 {
+		return nil, &dcrjson.RPCError{
+			Code:    dcrjson.ErrRPCInvalidParameter,
+			Message: "Invalid parameter, count must be > 0",
+		}
 	}
 
 	sv, err := s.chain.GetStakeVersions(hash, c.Count)

